### ISSUE TITLE
Unclear how to adapt multiple Keras preprocessing layers

### ISSUE THREAD

**Issue by tibbe** [tensorflow/tensorflow#50820]
## URL(s) with the issue:

https://www.tensorflow.org/guide/keras/preprocessing_layers

### Clear description

Currently all examples of calling `adapt()` on a preprocessing layer with a `Dataset` argument assumes a single input feature (and thus a single `Input`). In a model with multiple `Input`s and `PreprocessingLayer`s it's unclear how to use `adapt()` together with a `Dataset`.

### Usage example

(Cross-posted to https://stackoverflow.com/questions/68426974/how-do-you-use-keras-preprocessing-normalization-layers-with-multi-input-models):

```
x_norm = preprocessing.Normalization()
y_norm = preprocessing.Normalization()

x = layers.Input(shape=(1,))
x = x_norm(x)
y = layers.Input(shape=(1,))
y = y_norm(y)
concated = layers.Concatenate()([x, y])
output = layers.Dense(1)(concated)
model = keras.Model(inputs=[x, y], outputs=output)
```

This does *not* work but should:

```
x_norm.adapt(dataset)
y_norm.adapt(dataset)
```

Workaround that involve manual iteration over the `Dataset` result in inefficient execution as `adapt()` manages its own `tf.function` internally. This results in warning for the manual approach:

```
normalization_layers = {
    'x': preprocessing.Normalization(),
    'y': preprocessing.Normalization(),
    }
for batch in dataset:
  for name, layer in normalization_layers.items():
    layer.adapt(batch[0][name])
```

**Comment by tibbe** [comment]
More generally, it seems like `adapt()` is in the wrong place. It seems like it should be part of the model, just like `fit()`, so all layers can be discovered and adapted together: `model.adapt(dataset)`

**Comment by tibbe** [comment]
https://www.tensorflow.org/recommenders/examples/featurization seems to be the best example of how to work with several inputs. Perhaps https://www.tensorflow.org/guide/keras/preprocessing_layers clarifying the approach (e.g. use `Dataset.map` repeatedly). Seems rather inefficient though as you need one pass over the data per feature instead of adapting them all in a single pass.

**Comment by peteboothroyd** [comment]
This class offers a similar interface to what you describe:

https://github.com/tensorflow/tensorflow/blob/79e9e45fa08931aaa6bc72124cfbe776e40e2cc0/tensorflow/python/keras/layers/preprocessing/preprocessing_stage.py#L98-L133

**Comment by Saduf2019** [comment]
@tibbe 
Please update as per above comment

**Comment by tibbe** [comment]
Works well. Could perhaps a somewhat more elaborate example where the inputs/outputs of the stage is connected to a model. I'd imagine it'd look something like:

```py
preproc_layer = preprocessing_stage.FunctionalPreprocessingStage(
    inputs, normalization_layers)
preproc_layer.adapt(train_dataset)

x = tf.keras.layers.Concatenate()(preproc_layer.outputs)
x = ...
output = layers.Dense(1)(x)
model = keras.Model(inputs=preproc_layer.inputs, outputs=output)
```

Feel free to close.

**Comment by Saduf2019** [comment]
@tibbe 
Thank you for your update moving this to closed status as resolved.

**Comment by tibbe** [comment]
I spoke too soon. While the `FunctionalPreprocessingState.adapt()` method runs without errors it doesn't seem to work. If I individually adapt layers like so:

```py
inputs = {}
normalization_layers = {}

for name in NUMERIC_FEATURES:
  inputs[name] = layers.Input(shape=(1,))
  normalization_layers[name] = preprocessing.Normalization()

for name, layer in normalization_layers.items():
  layer.adapt(train_dataset.map(lambda x, y: x[name]))

input_layers = []
for name, input in inputs.items():
  input_layers.append(normalization_layers.get(name)(input))
x = tf.keras.layers.Concatenate()(input_layers)
x = layers.Dense(32)(x)
x = layers.Dense(16)(x)
output = layers.Dense(1)(x)
model = keras.Model(inputs=inputs, outputs=output)
model.compile(optimizer='adam', loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError(name='rmse')])
```

the model trains. If I instead use the helper class like so:

```py
inputs = []
normalization_layers = []

for name in NUMERIC_FEATURES:
  input = layers.Input(shape=(1,))
  inputs.append(input)
  normalization_layers.append(preprocessing.Normalization()(input))

preproc_layer = preprocessing_stage.FunctionalPreprocessingStage(
    inputs, normalization_layers)
preproc_layer.adapt(train_dataset)

x = tf.keras.layers.Concatenate()(preproc_layer.outputs)
x = layers.Dense(32, activation="relu")(x)
x = layers.Dense(16, activation="relu")(x)
output = layers.Dense(1)(x)
model = keras.Model(inputs=preproc_layer.inputs, outputs=output)
model.compile(optimizer='adam', loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError(name='rmse')])
```

It doesn't train (the loss isn't decreasing). I suspect that the layers didn't get adapted properly and are doing random transformations to the input features during training.

**Comment by tibbe** [comment]
I think this issue needs to be re-opened.

**Comment by Saduf2019** [comment]
@tibbe 

Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more refer to:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999

**Comment by tibbe** [comment]
https://github.com/keras-team/keras/issues/15219

**Comment by tibbe** [comment]
So the Keras people says (see previous comment for link to the Keras bug) that `FunctionalPreprocessingStage` isn't an exposed symbol of Keras (but I believe it is of TF as of 2.6) so it's now unclear to me who feels responsible for that function working. :)
